### PR TITLE
Decode PublicKeyCredentialOptions

### PR DESCRIPTION
**Pull request by ErinvanderVeen**
A small commit to Decode the PublicKeyCredentialOptions.

### CODE REVIEW

**Comment by Infinisil** [comment]
We should insert links and text of the spec where the `MUST ignore unknown` values are done.

File: fido/Crypto/Fido2/Model/JavaScript/Decoding.hs [hunk 1]
```diff
@@ -277,6 +284,123 @@ instance Decode (M.PublicKeyCredential 'M.Get) where
...
+        let pkcdTyp = M.PublicKeyCredentialTypePublicKey
+        pkcdId <- decode id
+        pkcdTransports <- decode transports
+        pure . Just $ M.PublicKeyCredentialDescriptor {..}
+      decodeDescriptor _ = pure Nothing
+
+instance Decode M.AuthenticatorAttachment where
+  decode "platform" = Right M.AuthenticatorAttachmentPlatform
+  decode "cross-platform" = Right M.AuthenticatorAttachmentCrossPlatform
+  decode attachment = Left $ DecodingErrorUnexpectedAttachment attachment
```
Comment (Infinisil) [inline fido/Crypto/Fido2/Model/JavaScript/Decoding.hs:339]: This shouldn't throw an error, see https://www.w3.org/TR/webauthn-2/#dom-authenticatorselectioncriteria-authenticatorattachment

File: fido/Crypto/Fido2/Model/JavaScript/Decoding.hs [hunk 2]
```diff
@@ -277,6 +284,123 @@ instance Decode (M.PublicKeyCredential 'M.Get) where
...
+
+instance Decode M.ResidentKeyRequirement where
+  decode "discouraged" = Right M.ResidentKeyRequirementDiscouraged
+  decode "preferred" = Right M.ResidentKeyRequirementPreferred
+  decode "required" = Right M.ResidentKeyRequirementRequired
+  decode requirement = Left $ DecodingErrorUnexpectedResidentKeyRequirement requirement
```
Comment (Infinisil) [inline fido/Crypto/Fido2/Model/JavaScript/Decoding.hs:345]: Same here, see https://www.w3.org/TR/webauthn-2/#dom-authenticatorselectioncriteria-residentkey

File: fido/Crypto/Fido2/Model/JavaScript/Decoding.hs [hunk 3]
```diff
@@ -277,6 +284,123 @@ instance Decode (M.PublicKeyCredential 'M.Get) where
...
+      decodeParam _ = pure Nothing
+
+instance Decode (M.PublicKeyCredentialOptions 'M.Create) where
+  decode JS.PublicKeyCredentialCreationOptions {..} = do
+    pkcocRp <- decode rp
+    pkcocUser <- decode user
+    pkcocChallenge <- decode challenge
+    pkcocPubKeyCredParams <- decode pubKeyCredParams
+    pkcocTimeout <- decode timeout
+    -- The default is []
```
Comment (Infinisil) [inline fido/Crypto/Fido2/Model/JavaScript/Decoding.hs:386]: Can be removed, same in the `'M.Get` instance

File: fido/Crypto/Fido2/Model/JavaScript/Decoding.hs [hunk 4]
```diff
@@ -277,6 +284,123 @@ instance Decode (M.PublicKeyCredential 'M.Get) where
...
+  decode JS.PublicKeyCredentialUserEntity {..} = do
+    pkcueId <- decode id
+    pkcueDisplayName <- decode displayName
+    pkcueName <- decode name
+    pure $ M.PublicKeyCredentialUserEntity {..}
+
+instance Decode M.Challenge
+
+instance Decode PublicKey.COSEAlgorithmIdentifier where
+  decode n = maybe (Left $ DecodingErrorUnexpectedAlgorithmIdentifier n) Right $ PublicKey.toAlg n
```
Comment (Infinisil) [inline fido/Crypto/Fido2/Model/JavaScript/Decoding.hs:311]: The spec doesn't seem to mention what should happen for unknown COSEAlgorithmIdentifier's, so I think erring out is good, but we should add a comment about this, explaining this, linking to the spec

File: fido/Crypto/Fido2/Model/JavaScript/Decoding.hs [hunk 5]
```diff
@@ -277,6 +284,123 @@ instance Decode (M.PublicKeyCredential 'M.Get) where
...
+instance Decode M.Timeout
+
+instance Decode [M.AuthenticatorTransport] where
+  decode = pure . mapMaybe decodeTransport
+    where
+      decodeTransport "usb" = Just M.AuthenticatorTransportUSB
+      decodeTransport "nfc" = Just M.AuthenticatorTransportNFC
+      decodeTransport "ble" = Just M.AuthenticatorTransportBLE
+      decodeTransport "internal" = Just M.AuthenticatorTransportInternal
+      decodeTransport _ = Nothing
```
Comment (Infinisil) [inline fido/Crypto/Fido2/Model/JavaScript/Decoding.hs:322]: All of these special cases should also have a link to the spec and an explanation

File: fido/Crypto/Fido2/Model/JavaScript/Decoding.hs [hunk 6]
```diff
@@ -277,6 +284,123 @@ instance Decode (M.PublicKeyCredential 'M.Get) where
...
+
+instance Decode [M.PublicKeyCredentialDescriptor] where
+  decode Nothing = pure []
+  decode (Just xs) = catMaybes <$> traverse decodeDescriptor xs
```
Comment (Infinisil) [inline fido/Crypto/Fido2/Model/JavaScript/Decoding.hs:326]: Here as well

File: fido/Crypto/Fido2/Model/JavaScript/Decoding.hs [hunk 7]
```diff
@@ -277,6 +284,123 @@ instance Decode (M.PublicKeyCredential 'M.Get) where
...
+  decode "preferred" = Right M.ResidentKeyRequirementPreferred
+  decode "required" = Right M.ResidentKeyRequirementRequired
+  decode requirement = Left $ DecodingErrorUnexpectedResidentKeyRequirement requirement
+
+instance Decode M.UserVerificationRequirement where
+  decode (Just "discouraged") = Right M.UserVerificationRequirementDiscouraged
+  decode (Just "preferred") = Right M.UserVerificationRequirementPreferred
+  decode (Just "required") = Right M.UserVerificationRequirementRequired
+  -- Default is "preferred"
+  decode _ = Right M.UserVerificationRequirementPreferred
```
Comment (Infinisil) [inline fido/Crypto/Fido2/Model/JavaScript/Decoding.hs:352]: Here as well, link to the spec

File: fido/Crypto/Fido2/Model/JavaScript/Decoding.hs [hunk 8]
```diff
@@ -277,6 +284,123 @@ instance Decode (M.PublicKeyCredential 'M.Get) where
...
+    ascUserVerification <- decode userVerification
+    pure $ M.AuthenticatorSelectionCriteria {..}
+
+instance Decode M.AttestationConveyancePreference where
+  decode (Just "none") = Right M.AttestationConveyancePreferenceNone
+  decode (Just "indirect") = Right M.AttestationConveyancePreferenceIndirect
+  decode (Just "direct") = Right M.AttestationConveyancePreferenceDirect
+  decode (Just "enterprise") = Right M.AttestationConveyancePreferenceEnterprise
+  -- Default is "none"
+  decode _ = Right M.AttestationConveyancePreferenceNone
```
Comment (Infinisil) [inline fido/Crypto/Fido2/Model/JavaScript/Decoding.hs:367]: Here as well

File: fido/Crypto/Fido2/Model/JavaScript/Decoding.hs [hunk 9]
```diff
@@ -277,6 +284,123 @@ instance Decode (M.PublicKeyCredential 'M.Get) where
...
+
+instance Decode [M.PublicKeyCredentialParameters] where
+  decode xs = catMaybes <$> traverse decodeParam xs
```
Comment (Infinisil) [inline fido/Crypto/Fido2/Model/JavaScript/Decoding.hs:370]: Same here

**Review by Infinisil — CHANGES_REQUESTED** [review]

**Comment by ErinvanderVeen** [comment]
Current failures are due `server/Main.hs` which is currently being reworked. @Infinisil Do you think it is worth fixing the server in this PR when it will be reworked soon anyway?

File: fido/Crypto/Fido2/Model/JavaScript/Decoding.hs [hunk 10]
```diff
@@ -80,6 +84,9 @@ data DecodingError
   | DecodingErrorNotAllInputUsed LBS.ByteString
   | DecodingErrorBinary String
   | DecodingErrorCBOR CBOR.DeserialiseFailure
+  | DecodingErrorUnexpectedAlgorithmIdentifier JS.COSEAlgorithmIdentifier
+  | DecodingErrorUnexpectedAttachment JS.DOMString
+  | DecodingErrorUnexpectedResidentKeyRequirement JS.DOMString
```
Comment (Infinisil) [inline fido/Crypto/Fido2/Model/JavaScript/Decoding.hs:89]: Two of these new errors aren't used anymore I believe

**Review by Infinisil — CHANGES_REQUESTED** [review]

File: fido/Crypto/Fido2/Model/JavaScript.hs [hunk 11]
```diff
@@ -217,6 +217,7 @@ data PublicKeyCredential response = PublicKeyCredential
     -- | [(spec)](https://www.w3.org/TR/webauthn-2/#dom-publickeycredential-response)
     response :: response,
     -- | [(spec)](https://www.w3.org/TR/webauthn-2/#dom-publickeycredential-getclientextensionresults)
+    -- TODO: Currently a Maybe because webauthn-json doesn't propagate this field to the server.
```
[the remark this answers is not in the document]
Comment (Infinisil) [reply fido/Crypto/Fido2/Model/JavaScript.hs:220]: https://github.com/github/webauthn-json/blob/main/src/basic/schema.ts#L62

[the remark this answers is not in the document]
Comment (ErinvanderVeen) [reply fido/Crypto/Fido2/Model/JavaScript.hs]: I've added them to the tests.

**Review by Infinisil — APPROVED** [review]
Looking good!

**Pull request merged**